server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java

```diff
@@ -114,6 +127,32 @@ class DateHistogramAggregator extends BucketsAggregator implements SizedBucketAg
         this.formatter = valuesSourceConfig.format();
 
         bucketOrds = LongKeyedBucketOrds.build(context.bigArrays(), cardinality);
+
+        // Create the filters for fast aggregation only if the query is instance
+        // of point range query and there aren't any parent/sub aggregations
+        if (parent() == null && subAggregators.length == 0) {
+            final String fieldName = valuesSourceConfig.fieldContext().field();
+            if (context.query() instanceof IndexOrDocValuesQuery) {
+                final IndexOrDocValuesQuery q = (IndexOrDocValuesQuery) context.query();
+                if (q.getIndexQuery() instanceof PointRangeQuery) {
+                    final PointRangeQuery prq = (PointRangeQuery) q.getIndexQuery();
+                    // Ensure that the query and aggregation are on the same field
+                    if (prq.getField().equals(fieldName)) {
+                        createFilterForAggregations(fieldName,
+                            NumericUtils.sortableBytesToLong(prq.getLowerPoint(), 0),
+                            NumericUtils.sortableBytesToLong(prq.getUpperPoint(), 0));
+                    }
+                }
+            } else if (context.query() instanceof ConstantScoreQuery) {
+                final ConstantScoreQuery csq = (ConstantScoreQuery) context.query();
+                // Ensure that the constant score query is instance of match all query
+                if (csq.getQuery() instanceof MatchAllDocsQuery) {
+                    findBoundsAndCreateFilters(fieldName, context);
+                }
+            } else if (context.query() instanceof MatchAllDocsQuery) {
+                findBoundsAndCreateFilters(fieldName, context);
+            }
+        }
```

Comment:
Can you factor this logic out into a `private static` helper method? In particular, I think the helper method could be recursive -- there's technically nothing stopping someone from wrapping queries multiple times (plus e.g. wrapping in a `FunctionScoreQuery`, say, would also be viable, though silly).

Also, maybe outside the scope of this specific issue, since you're operating at the full shard level (assuming I'm following it correctly), but we could also apply this optimization to segments where the top-level query is "functionally" match-all (i.e. the top-level query is something other than these special cases, but happens to match all docs in the segment).

Reply:
Thank you @msfroh for the suggestion. I was also thinking on writing this logic better, guess the missing keyword was `recursion`. I will write this logic recursively for unwrapping the underlying concrete query, could be `PointRangeQuery` or `MatchAll`.

Reply:
> Also, maybe outside the scope of this specific issue, since you're operating at the full shard level (assuming I'm following it correctly), but we could also apply this optimization to segments where the top-level query is "functionally" match-all (i.e. the top-level query is something other than these special cases, but happens to match all docs in the segment).

The constructor logic operates at the full shard level. `getLeafCollector` is invoked once per segment. The only challenge I see is that the leaf level collector does not have direct visibility into the leaf level query. Although, these are the type of cases that we can gradually expand towards.

Reply:
> The only challenge I see is that the leaf level collector does not have direct visibility into the leaf level query.

I believe we can handle that with the `setWeight` to hold onto the top-level query's `Weight`. Then in `getLeafCollector`, we can use `weight.count(leafReaderContext)` to see if `count == numDocs`. That would mean that the top-level query is functionally a match-all.

That said, I don't think we necessarily need to do that right now -- I'm just excited about opportunities to use the `setWeight` method. (See https://github.com/opensearch-project/OpenSearch/issues/10954.)